### PR TITLE
Support F-Client and replace non-extended extension messages

### PR DESCRIPTION
**Pull request by heinrich5991**
This continues what #3439 started, and also replaces non-extended messages that are DDRace extensions with extended messages. The DDNet server still sends the old message to old clients and the DDNet client still understands the old messages.

Supersedes #3439.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
What do you think? @fokkonaut @Learath2

**Comment by fokkonaut** [comment]
Exactly what I tried to avoid: Using the normal playerflags. 0.7 is still in development, and a new playerflag can always come which would break compability. Also this means I need to change the handling of unpacking the playerflags in the client, because they will fail on unpacking them.

**Comment by fokkonaut** [comment]
Also the netobj handler 7 of ddnet server cannot unpack the playerflags.

**Comment by fokkonaut** [comment]
Exactly what I tried to avoid: Using the normal playerflags. 0.7 is still in development, and a new playerflag can always come which would break compability.

**Comment by Learath2** [comment]
Is `Sv_DDRaceTime` used for anything on the client? Do we really need to keep support for prehistoric clients? If yes maybe we can just remove the msg for new clients all together?

I like the change, a rudimentary look suggests it's fine, but I need to give it a more thorough read.

**Comment by heinrich5991** [comment]
Yes, `Sv_DDRaceTime` is used in the client.

File: src/game/server/player.cpp [hunk 1]
```diff
@@ -353,6 +355,8 @@ void CPlayer::Snap(int SnappingClient)
 			return;
 
 		pPlayerInfo->m_PlayerFlags = PlayerFlags_SixToSeven(m_PlayerFlags);
+		if(ClientVersion >= VERSION_DDRACE && (m_PlayerFlags & PLAYERFLAG_AIM))
```
Comment (fokkonaut) [inline src/game/server/player.cpp:358]: This needs to check for the client version of the snapping client, not the own one.

**Comment by fokkonaut** [comment]
I adapted to these changes and on my own server it works, just not correctly with this commit applied to ddnet servers.
F-Client commits: https://github.com/fokkonaut/F-Client/commit/a972c73d2404240ffe57737d7dabc2db03fe1b0f, https://github.com/fokkonaut/F-Client/commit/8ecd8c7f0ab4b41f16d165bc5ec9b9e87578012c
Also I noticed projectiles are invisible due to the extra info thats not used in 0.7 and therefore breaks the projectile. So we would need a check there too. I can receive teams, but not aimlines, seems like the ddnet server is not getting the clientver, in my mod it does get it.

**Comment by fokkonaut** [comment]
Ah, I found why the hookline wont work. F-Client is not sending it because you need to increase the `SERVERCAP_CURVERSION` to 2 in protocol_ex.h. But what do we do about the projectile check? Could you add a !IsSixup there? (I hope it wont break anything, because F-Client is sending the latest ddnet version right now)

File: src/game/server/player.cpp (hunk 1 above)
Comment (heinrich5991) [reply src/game/server/player.cpp]: Done, I think.

**Comment by heinrich5991** [comment]
> Ah, I found why the hookline wont work. F-Client is not sending it because you need to increase the `SERVERCAP_CURVERSION` to 2 in protocol_ex.h.

Done. Thanks for figuring that out.

> But what do we do about the projectile check? Could you add a !IsSixup there? (I hope it wont break anything, because F-Client is sending the latest ddnet version right now)

I don't see a difference between the 0.6 and 0.7 projectile netobj. Can you point me to the difference or where I should put that `!IsSixup` check?

**Comment by fokkonaut** [comment]
https://github.com/ddnet/ddnet/blob/5d45f061fabf0866a9e8e6d5e59d542d3188ca78/src/game/server/entities/projectile.cpp#L325

Here. I hope there are no other cases in which the DDNet client version nr is passing F-Client through some stuff it doesnt have. We would need ClientCaps for the server :D

**Comment by heinrich5991** [comment]
@fokkonaut What do you think? I can PR support for DDNETPROJECTILE to F-Client if you want.

**Comment by fokkonaut** [comment]
Sure.

File: src/game/server/entities/projectile.cpp [hunk 2]
```diff
@@ -319,13 +320,27 @@ void CProjectile::Snap(int SnappingClient)
...
-			FillExtraInfo(pProj);
-		else
-			FillInfo(pProj);
+		int Type = SnappingClientVersion <= VERSION_DDNET_MSG_LEGACY ? (int)NETOBJTYPE_PROJECTILE : NETOBJTYPE_DDNETPROJECTILE;
+		void *pProj = Server()->SnapNewItem(Type, GetID(), sizeof(DDNetProjectile));
+		if(!pProj)
+		{
+			return;
+		}
+		mem_copy(pProj, &DDNetProjectile, sizeof(DDNetProjectile));
```
Comment (Learath2) [inline src/game/server/entities/projectile.cpp:334]: This assumes the legacy and the new projectile message are laid out the same way in memory. Is this a fair assumption to make now? Is it guaranteed to stay a fair assumption in the future?

At least we should document this coupling somewhere

**Review by Learath2 — COMMENTED** [review]
We seem to be sending both the legacy and the new message to clients that seem new enough to receive the new message. Isn't that a bit wasteful (though convenient)?

**Comment by heinrich5991** [comment]
> We seem to be sending both the legacy and the new message to clients that seem new enough to receive the new message. Isn't that a bit wasteful (though convenient)?

I intended to always send the new message. And when the client is old, also send the legacy message it understands. This wastes network traffic for old clients, but eventually™ everyone upgrades to a client that is new enough.

File: src/game/server/entities/projectile.cpp (hunk 2 above)
Comment (heinrich5991) [reply src/game/server/entities/projectile.cpp:334]: I added documentation that it needs to have the same layout as `Projectile`. We already assume that we can read the structs as `int` arrays in the snapshot code.

**Comment by Learath2** [comment]
LGTM

bors r+

**Pull request merged**